[DI - Step 1] 러너덕 미션 제출합니다.

Deocksoo opened the pull request:
안녕하세요 후니! 이번 미션 리뷰이 러너덕입니다 :)
이번 미션을 진행하면서 궁금했던 사항들이 몇 가지 있었는데요, 리뷰어님께 의견 구하고 싶어서 일단 리뷰 요청 드려요, 의문사항은 아래와 같습니다.

1. BeanFactory 클래스 재귀 함수
BeanFactory 클래스의 initializeInjectedBean2() 메서드는 요구사항을 만족시키기 위해 처음 작성한 재귀 함수입니다. 현재 저희는 initializeInjectedBean() 메서드와  같이 재귀 함수를 조금 더 분리해서 리펙토링을 진행해 보았는데요, 이와 같이 리펙토링하면  한 메서드 안에서 재귀 호출이 일어나는 것이 아니라, 여러 메서드를 건너간 후에야 재귀 호출이 일어나 이해하기 어렵지 않을까 하는 생각이 들었습니다. 혹시 initializeInjectedBean()와 같이 분리하는 것이 좋을지, initializeInjectedBean2() 처럼 그대로 두는 것이 좋을지, 아니면 다른 더 좋은 방법이 있을지, 리뷰어님의 의견이 궁금합니다!

2. BeanFactory 클래스 assert 구문
inject() 메서드에서 injectedConstructor 변수에 대한 NULL체크를 assert를 통해 하고 있습니다. 그런데 이전에 호출되는 함수인 initializeNotInjectedBeans() 메서드에서 이미 NULL체크를 하고 있습니다. 지금과 같은 코드에서는 NULL체크를 두 번 하고 있는데, 이렇게 중복해서 체크하더라도 명시적으로 assert를 통해 체크하는 것이 좋은지, 리뷰어님의 의견이 궁금합니다!

다른 부분에서도 부족한 부분이 있으면 피드백 부탁드려요 :)
읽어주셔서 감사합니다, 잘 부탁드립니다 ㅎㅎ

pobiconan committed:
apply slingeton pattern to UserDao, JdbcTemplate

pobiconan committed:
refactor: jdbc url

pkch93 committed:
docs: 요구사항 정리

pkch93 committed:
feat: BeanFactory initialize 메서드 구현

pkch93 committed:
refactor: BeanFactory.initialize() 메서드 리팩토링

pkch93 committed:
docs: 의문사항 작성

pkch93 committed:
refactor: typo 수정

pkch93 committed:
docs: todo 추가

pkch93 committed:
feat: BeanScanner 구현

pkch93 committed:
refactor: AnnotationHandlerMapping에 BeanFactory 주입

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -20,17 +21,16 @@
 
-    public AnnotationHandlerMapping(Object... basePackage) {
-        this.basePackage = basePackage;
+    public AnnotationHandlerMapping(BeanFactory beanFactory) {
```
👍

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -20,17 +21,16 @@
     public void initialize() {
-        ControllerScanner controllerScanner = new ControllerScanner(basePackage);
-        Map<Class<?>, Object> controllers = controllerScanner.getControllers();
+        Map<Class<?>, Object> controllers = beanFactory.getControllers();
```
Map을 사용할 필요는 없을 것 같아요.

pkch93 committed:
refactor: BeanFactory inject, getInstance 메서드명 변경

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -20,17 +21,16 @@
-        Map<Class<?>, Object> controllers = controllerScanner.getControllers();
+        Map<Class<?>, Object> controllers = beanFactory.getControllers();
         Set<Method> methods = getRequestMappingMethods(controllers.keySet());
         for (Method method : methods) {
```
controllers.get(method.getDeclaringClass())는 BeanFactory에서 꺼내올 수 있을 것 같아요.

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
+
+    @SuppressWarnings("unchecked")
     public void initialize() {
+        initializeNotInjectedBeans();
```
결국엔 중복되는 로직이 발생할 것 같은데 이렇게 두가지로 분리한 이유가 있나요?

Deocksoo replied:
앗 이 부분은 중복이 있어서 수정하였습니다!

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
+        return parameters;
+    }
+
+    private void initializeInjectedBean2(Class clazz) {
```
해당 메서드에서는 너무 많은 일을 하고 있는 것 같아요.
분리 잘하셨네요 👍

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
 
+    private Object inject(Class concreteClass) {
+        Constructor<?> injectedConstructor = BeanFactoryUtils.getInjectedConstructor(concreteClass);
+        assert injectedConstructor != null;
```
어떤 exception이 날아가나요?
또 해당 exception을 어떻게 처리하나요?

Deocksoo replied:
이 부분도 중복을 수정하는 과정에서 자연스럽게 사라졌어요 :)
만약 assert를 그대로 사용하였더라면 AsserstionError exception이 날아갈테고, 해당 exception 처리를 어디에선가 추가로 해주었어야 했겠네요. 이미 중복체크를 앞에서 하고 있다고 생각해서 그 부분은 고려하지 못했는데, 만약 이런 방식으로 계속 구현했더라면 생각했어야 하는 부분인 것 같아요.
일단은 저희가 리펙토링 한 방향으로 수정하여 한번 pr 보내드리겠습니다!

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
 import java.util.Set;
+import java.util.stream.Collectors;
 
 public class BeanFactory {
```
중복되는 부분들이 정리가 되면 좋을 것 같아요 ㅠㅠ
코드 정리를 해주세요!

pkch93 committed:
refactor: BeanFactory initialize 메서드 중복제거

- initializeNotInjectedBeans 메서드가 initializeInjectedBeans와 같은 기능을
수행하고 있으므로 중복 부분 제거

jeonghoon1107 reviewed (changes requested):
안녕하세요 러너덕!
구현 잘 하셨네요 👍 
몇가지 피드백 남겼으니 확인부탁드려요 :)

pkch93 committed:
refactor: test 코드 package 구조 변경

pkch93 committed:
refactor: BeanScannerTest의 scanAll 메서드 DisplayName 추가

pkch93 committed:
docs: 추가 테스트 목록 작성

pkch93 committed:
test: BeanScanner 테스트케이스 추가

pkch93 committed:
docs: 의문사항 추가

pkch93 committed:
refactor: BeanFactoryUtils getInjectedConstructor 메서드

1. 테스트 케이스를 추가하였습니다.
2. Injected 생성자가 여러 개일 경우 예외를 던지도록 수정하였습니다.

pkch93 committed:
refactor: BeanFactoryUtils getInjectedConstructor 메서드

1. 테스트 케이스를 추가하였습니다.
2. Injected 생성자가 여러 개일 경우 예외를 던지도록 수정하였습니다.

pkch93 committed:
refactor: di 프레임워크 적용하여 Controller, Dao 리펙토링

pkch93 committed:
refactor: AnnotationHandlerMapping

- addHandlerExecutions 메서드 불필요한 인자 제거
- BeanFactory.getControllers 메서드가 Set 반환하도록 변경

pkch93 committed:
refactor: BeanFactory.initialize 메서드에서 불필요한 bypass 제거

pkch93 committed:
test: BeanFactory

- 생성자 매개변수로 같은 종류의 어노테이션을 가진 빈을 생성하는 테스트 구현

Deocksoo commented:
후니님 안녕하세요, 신경써 리뷰해주셔서 감사합니다:)
어제 피드백 때 말씀해주셨던 부분들은 모두 반영하는 것이 좋을것 같아서 수정해보았어요!
더불어 어제 코멘트로 여쭤봐주셨던 것들에 대해서도 간단한 코멘트를 남겨두었습니다.
피드백 부탁드립니다 :)

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,102 @@
+        try {
+            return constructor.newInstance(parameters);
+        } catch (IllegalAccessException | InstantiationException | InvocationTargetException e) {
+            logger.error(e.getMessage(), e);
```
에러 메시지가 어떻게 찍혔는지 확인해보셨나요?
e.getMessage()말고
logger.error("추적하기 쉬운 메시지", e);
이렇게 남겨도 좋을 것 같아요!

Deocksoo replied:
이 부분은 습관적으로 이렇게 작성하고 있었는데, 생각해보니 catch한 에러 메시지가 보기 어렵게 되어있을 수 있겠네요. step2 시작 전에 반영해두겠습니다!

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactoryUtils.java:
```diff
@@ -20,10 +20,15 @@
-            return null;
+        validateMultipleInjectedConstructor(injectedConstructors);
+
+        return injectedConstructors.isEmpty() ? null : injectedConstructors.iterator().next();
```
해당 메서드에서 null을 반환해주고 사용하는 곳에서 다시 null check를 하는 행위를 다시 생각해보면 좋을 것 같아요!

Deocksoo replied:
말씀해주신 부분을 조금 더 고민해봐야 할거같아요. 다시 보니 확실히 이상하네요.. 어떤 방법이 있을지 조금 더 고민해보고 다음 pr에 반영하겠습니다!

jeonghoon1107 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactoryUtils.java:
```diff
@@ -20,10 +20,15 @@
         Set<Constructor> injectedConstructors = getAllConstructors(clazz, withAnnotation(Inject.class));
-        if (injectedConstructors.isEmpty()) {
-            return null;
+        validateMultipleInjectedConstructor(injectedConstructors);
```
👍

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -60,9 +60,9 @@ private void addHandlerExecutions(Map<Class<?>, Object> controllers, Method meth
 
     @SuppressWarnings("unchecked")
-    private Set<Method> getRequestMappingMethods(Set<Class<?>> controlleers) {
+    private Set<Method> getRequestMappingMethods(Set<Class<?>> controllers) {
```
annotation을 함께 인자로 받으면 좀 더 범용적으로 사용하면 어떨까요?
특정 클래스의 특정 annotation이 달린 메서드를 가져오는 메서드라던지..
Util 성 클래스로 빼면 AnnotationHandlerMapping 클래스가 더 깔끔해질 것 같아요!

Deocksoo replied:
아하.. 그렇겠네요! 좋은 생각인 것 같아요!
한번 수정해보겠습니다, 그러면 역할이 조금 더 깔끔하게 나뉠 수 있을것 같아요!

jeonghoon1107 reviewed (approved):
안녕하세요 러너덕!
리팩토링 깔끔하게 잘 하셨네요 👍 
다음 단계 진행해주세요 :)
